### PR TITLE
HDDS-9959. Propagate group remove to other datanodes during pipeline close

### PR DESCRIPTION
**Pull request by ivandika3**
## What changes were proposed in this pull request?

In https://issues.apache.org/jira/browse/RATIS-1947, it was found that there might be cases where RaftServer for Datanodes in the same pipeline are closed hours apart. 

```
# dn1
2023-11-29 15:22:59,477 [Command processor thread] INFO org.apache.hadoop.ozone.container.common.statemachine.commandhandler.ClosePipelineCommandHandler: Close Pipeline PipelineID=23e46782-6b48-4559-b3ac-0f95993cf0bc command on datanode 1669a7e6-fe3c-4f7e-8fcb-ec5d5027b0eb.


#dn5
2023-11-29 14:07:55,442 [Command processor thread] INFO org.apache.hadoop.ozone.container.common.statemachine.commandhandler.ClosePipelineCommandHandler: Close Pipeline PipelineID=23e46782-6b48-4559-b3ac-0f95993cf0bc command on datanode bd1e72ab-cfd5-4cc1-8fbf-6ec9d9654c98.


# dn8
2023-11-29 16:57:53,894 [Command processor thread] INFO org.apache.hadoop.ozone.container.common.statemachine.commandhandler.ClosePipelineCommandHandler: Close Pipeline PipelineID=23e46782-6b48-4559-b3ac-0f95993cf0bc command on datanode 4a23d1e8-d526-4a4d-8ed1-13ffbab3a5cc.
This might happen when there are a lot of commands in some of the Datanode commandQueue, causing some pipeline close commands to be handled by datanodes later than the others.
```

Furthermore, Ratis group remove operation is local to the Raft server and is not propagated to the other Raft peers in the same group. Therefore, datanodes that have not received the group remove operation will keep operating (e.g. sending RequestVote / AppendEntries RPCs), although the pipeline (Raft group) is supposed to be closed. The might affect the client communicating with the Raft peer in datanodes that have not been closed yet.

Therefore, similar to CreatePipelineCommandHandler, the first datanode that receives the close pipeline command needs to propagate the group remove command to the other datanodes (Raft peers) in the same pipeline. This will close the pipeline immediately on all the datanodes. The subsequent close pipeline commands will be ignored silently as the pipeline has been successfully closed. 

Note: There seems to be an unused methods in `RatisPipelineUtils` related to pipeline destroy (`RatisPipelineUtils#destroyPipeline`) which were re-introduced in https://github.com/apache/ozone/pull/538 although they were not really used. However, it seems the intent of the methods are for the SCM to destroy the pipeline directly, without going through the command queue. We can delete in this patch if needed.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9959

## How was this patch tested?

Unit test for `ClosePipelineCommandHandler`.

CI run: https://github.com/ivandika3/ozone/actions/runs/7258608368/job/19774884832

### CODE REVIEW

**Comment by ivandika3** [comment]
@sumitagrawl @DaveTeng0 @szetszwo Could you help to review if you have time?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ClosePipelineCommandHandler.java [hunk 1]
```diff
@@ -84,6 +106,27 @@ public void handle(SCMCommand command, OzoneContainer ozoneContainer,
...
+                .filter(peer -> peer.getId() != ratisServer.getServer().getId())
+                .forEach(peer -> {
+                  try (RaftClient client = newRaftClient.apply(peer, ozoneContainer.getTlsClientConfig())) {
+                    client.getGroupManagementApi(peer.getId())
+                        .remove(raftGroupId, shouldDeleteRatisLogDirectory, !shouldDeleteRatisLogDirectory);
+                  } catch (GroupMismatchException ae) {
+                    // ignore silently since this means that the group has been closed by earlier close pipeline
+                    // command in another datanode
+                  } catch (IOException ioe) {
+                    LOG.warn("Remove group failed for peer {}", peer.getId(), ioe);
```
Comment (szetszwo) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ClosePipelineCommandHandler.java:126]: Let's include the group id.
```java
                    LOG.warn("Failed to remove group {} for peer {}", raftGroupId, peer.getId(), ioe);
```

**Review by szetszwo — COMMENTED** [review]
@ivandika3 , thanks for working on this!  Please see the comments inlined.

> ... There seems to be an unused methods in RatisPipelineUtils related to pipeline destroy (RatisPipelineUtils#destroyPipeline) ...

Sure, let's remove the unused methods.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ClosePipelineCommandHandler.java [hunk 2]
```diff
@@ -84,6 +106,27 @@ public void handle(SCMCommand command, OzoneContainer ozoneContainer,
...
           server.removeGroup(pipelineIdProto);
+          if (server instanceof XceiverServerRatis) {
+            // TODO: Refactor Ratis logic to XceiverServerRatis
+            // Propagate the group remove to the other Raft peers in the pipeline
+            XceiverServerRatis ratisServer = (XceiverServerRatis) server;
+            final RaftGroupId raftGroupId = RaftGroupId.valueOf(pipelineID.getId());
+            final Collection<RaftPeer> peers = ratisServer.getRaftPeersInPipeline(pipelineID);
+            final boolean shouldDeleteRatisLogDirectory = ratisServer.getShouldDeleteRatisLogDirectory();
+            peers.stream()
+                .filter(peer -> peer.getId() != ratisServer.getServer().getId())
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ClosePipelineCommandHandler.java]: Updated, thank you for catching it.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ClosePipelineCommandHandler.java (hunk 1 above)
Comment (ivandika3) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/ClosePipelineCommandHandler.java]: Updated. PTAL.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by adoroszlai** [comment]
Test failure in `TestBlockOutputStreamWithFailures.test2DatanodesFailure` seems related, and it has not been flaky before.

https://github.com/apache/ozone/actions/runs/7283278738/job/19847237481?pr=5827#step:6:132

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Test cases to verify ClosePipelineCommandHandler.
+ */
+@ExtendWith(MockitoExtension.class)
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java:54]: `MockitoExtension` is unnecessary, there are no variables annotated as `@Mock`.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  private StateContext stateContext;
+  private SCMConnectionManager connectionManager;
+  private RaftClient raftClient;
+  private GroupManagementApi raftClientGroupManager;
+  private OzoneConfiguration conf;
+
+  @BeforeEach
+  public void setup() throws Exception {
+    conf = new OzoneConfiguration();
+    ozoneContainer = Mockito.mock(OzoneContainer.class);
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java:67]: nit: Please add static import for `mock()`, `when()`, `any()`, etc.

**Review by adoroszlai — COMMENTED** [review]
Thanks @ivandika3 for the patch.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java (hunk 3 above)
Comment (ivandika3) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java]: Thank you for the review. Updated.

We might also need to remove this for `TestCreatePipelineCommandHandler`.

File: hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java (hunk 4 above)
Comment (ivandika3) [reply hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/TestClosePipelineCommandHandler.java]: Thank you for the review. Updated.

**Comment by adoroszlai** [comment]
Thanks @ivandika3 for the patch, @szetszwo for the review.

**Comment by ivandika3** [comment]
@szetszwo Thanks for the info. Let me take a look.

**Pull request merged**